rpc/server.go

```diff
@@ -0,0 +1,364 @@
+	}
+	return rpcCtx
+}
+
+// processSubscriptionRequest handles the incoming subscription request
+func (r *RpcServer) processSubscriptionRequest(req *types.RPCRequestBody) (*handler, error) {
+	// subscription method is the first argument
+	method, err := parseSubscriptionMethod(req.Params.Value)
+	if err != nil {
+		return nil, &errs.InvalidParamsError{Message: err.Error()}
+	}
+
+	r.mu.RLock()
+	s, ok := r.serviceMap.subcriptions[subscriptionPrefix+strings.ToLower(method)]
+	r.mu.RUnlock()
+	if !ok {
+		return nil, &errs.SubscriptionNotFoundError{Subscription: method}
+	}
+	return s.handler, nil
+}
+
+// callSubscribe processes *_subscribe methods
+func (r *RpcServer) callSubscribe(ctx *context.Context, rpcCtx *RpcContext) *RpcContext {
+	req := rpcCtx.parsedBody
+	handler, err := r.processSubscriptionRequest(req)
+	if err != nil {
+		return rpcCtx.SetErrorObject(&errs.GenericError{Err: err})
+	}
+
+	// Parse subscription name arg too, but remove it before calling the callback.
+	argTypes := append([]reflect.Type{stringType}, handler.argTypes...)
+	args, err := prepareArguments(req.Params.Value, argTypes)
+	if err != nil {
+		return rpcCtx.SetErrorObject(&errs.InvalidParamsError{Message: err.Error()})
+	}
+	args = args[1:]
+
+	// Add notifier to context so that subscription handler can use it
+	n := &Notifier{h: handler, wsCtx: rpcCtx.wsCtx}
+	r.notifiers = append(r.notifiers, n)
+
+	*ctx = PutNotifierKey(*ctx, n)
+	resp, err := handler.call(*ctx, args)
+	if err != nil {
+		return rpcCtx.SetErrorObject(&errs.GenericError{Err: err})
+	}
+
+	respJson, err := jsoniter.Marshal(resp)
+	if err != nil {
+		return rpcCtx.SetErrorObject(&errs.GenericError{Err: err})
+	}
+
+	return rpcCtx.setResult(respJson)
+}
+
+// callUnsubscribe processes *_unsubscribe methods
+func (r *RpcServer) callUnsubscribe(rpcCtx *RpcContext) *RpcContext {
+	subscriptionId, subscriptionIdDataType, _, err := jsonparser.Get(rpcCtx.parsedBody.Params.Value, "[0]")
+	if err != nil {
+		return rpcCtx.SetErrorObject(&errs.InvalidParamsError{Message: fmt.Sprintf("can't parse eth_unsibscribe params: %v", err)})
+	}
+	if subscriptionIdDataType != jsonparser.String {
+		return rpcCtx.SetErrorObject(&errs.InvalidParamsError{Message: "subscription ID must be string"})
+	}
+
+	// scan the notifiers to find and remove the subscription
+	for i := range r.notifiers {
+		if r.notifiers[i].sub.ID == ID(subscriptionId) {
+			close(r.notifiers[i].sub.err)
+			// remove the wsCtx.subscription used by the CloseWsConn method
+			delete(r.notifiers[i].wsCtx.subscriptions, ID(subscriptionId))
+			return rpcCtx.setResult([]byte("true"))
+		}
+	}
+	return rpcCtx.setResult([]byte("false"))
+}
+
+// callMethod processes incoming service methods
+func (r *RpcServer) callMethod(ctx *context.Context, rpcCtx *RpcContext) *RpcContext {
+	r.mu.RLock()
+	s, ok := r.serviceMap.services[strings.ToLower(rpcCtx.parsedBody.Method.Str())]
+	r.mu.RUnlock()
```

Comment:
The aim of this lock was not to block service registration because of intensive read access. Each callMethod works in a separate go routine since it is called from http server but I may be missing something. I would be happy to discuss further if you have any concerns.